woocommerce/woocommerce
# Fix issue with empty table row added to add_to_cart widget

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Since version 9.5.0 the html markup for the variations table in the add_to_cart widget has been changed to move the evaluation for the Clear button in a dedicated table row. 

The previous code looked like this
```
<td class="value">
    <?php
        wc_dropdown_variation_attribute_options(
            array(
                'options'   => $options,
                'attribute' => $attribute_name,
                'product'   => $product,
            )
        );
        echo end( $attribute_keys ) === $attribute_name ? wp_kses_post( apply_filters( 'woocommerce_reset_variations_link', '<a class="reset_variations" href="#">' . esc_html__( 'Clear', 'woocommerce' ) . '</a>' ) ) : '';
    ?>
</td>
```

Now this has been moved in its own table row, as follow
```
<tr>
    <td colspan="2">
        <?php
        /**
         * Filters the reset variation button.
         *
         * @since 2.5.0
         *
         * @param string  $button The reset variation button HTML.
         */
        echo end( $attribute_keys ) === $attribute_name ? wp_kses_post( apply_filters( 'woocommerce_reset_variations_link', '<button class="reset_variations"  aria-label="' . esc_html__( 'Clear options', 'woocommerce' ) . '">' . esc_html__( 'Clear', 'woocommerce' ) . '</button>' ) ) : '';
        ?>
    </td>
</tr>
```

The consequence of this change is that when the attribute being rendered is not the latest, an empty row is added to table below the attribute row, as you can see in this screenshot

![Screenshot 2024-12-18 165013](https://github.com/user-attachments/assets/9541b1cc-c348-4b65-91b3-1deddeb6c382)

This PR  changes the markup to include that table row only when 
1. The `woocommerce_reset_variations_link` filter returns some content.
2. When the attribute being rendered is the latest.

The problem with the current solution is that that row cannot be removed in any other way than overwriting the template. 
This  PR solution ensure that whatever choice has been made in the `woocommerce_reset_variations_link` filter (like show no button) is respected and no uneccessary empty markup is added to the page.

### Testing steps

1. You need a product with 2 variation attributes to select from.
2. Go to the product page, notice that there is no empty space/cell between the first and second variation rows (you can inspect if you want).
3. Select variations, make sure the clear button is rendered correctly.

### Changelog entry












<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 
Fix issues with empty table rows added after each attribute in the add_to_cart widget when the attribute being rendered was not the latest.
</details>
<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
